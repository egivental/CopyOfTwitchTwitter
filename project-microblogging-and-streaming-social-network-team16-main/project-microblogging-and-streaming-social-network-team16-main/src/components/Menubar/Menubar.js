import React from "react";
import { useSelector } from "react-redux";
import Navbar from "react-bootstrap/Navbar";
import Nav from "react-bootstrap/Nav";

/**
 * @returns {Menubar} A Menubar React component.
 */
function Menubar() {
  const userState = useSelector((state) => state.user);
  let navLinks = (
    <Nav className="ml-auto">
      <Nav.Link href="/signup">Sign Up</Nav.Link>
      <Nav.Link href="/signin">Sign In</Nav.Link>
    </Nav>
  );
  if (userState.isUserLoggedIn) {
    navLinks = (
      <Nav className="ml-auto">
        <Nav.Link href="/feed">Posts</Nav.Link>
        <Nav.Link href="/streams">Livestreams</Nav.Link>
        <Nav.Link href="/contacts">Contacts</Nav.Link>
        <Nav.Link href="/inbox">Messages</Nav.Link>
        <Nav.Link href="/profile">Profile</Nav.Link>
        <Nav.Link href="/signout">Sign Out</Nav.Link>
      </Nav>
    );
  }

  return (
    <Navbar bg="light" expand="lg">
      <Navbar.Brand href="/home">
        <strong>ABE Twitter</strong>
      </Navbar.Brand>
      <Navbar.Toggle />
      <Navbar.Collapse id="basic-navbar-nav">{navLinks}</Navbar.Collapse>
    </Navbar>
  );
}

export default Menubar;
