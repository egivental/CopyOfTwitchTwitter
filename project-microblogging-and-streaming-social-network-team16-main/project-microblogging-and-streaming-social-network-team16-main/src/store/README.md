A short guide on how to use the store in this folder:

## Imports
```
import React, { useState } from "react";
import { useSelector, useDispatch } from "react-redux";
```

## Local State in a Component
```
const [incrementAmount, setIncrementAmount] = useState("2");

```

where `incrementAmount` is the getter and `setIncrementAmount()` is the setter. Use `useState("2")` makes the default value of `incrementAmount` to be `2`.

## Create a Store State
`counterState.js`
```
import { createSlice } from "@reduxjs/toolkit";

export default createSlice({
  name: "counter",
  initialState: {
    value: 0,
  },
  reducers: {
    increment: (state) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the Immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based on those changes
      state.value += 1;
    },
    decrement: (state) => {
      state.value -= 1;
    },
    incrementByAmount: (state, action) => {
      state.value += action.payload;
    },
  },
});

```

## Import the Counter State
```
import counterState from "./counterState";
```

## Read Store State in a Component
```
const count = useSelector((state) => state.counter.value);
```

## Perform Actions on a Store State in a Component
```
const dispatch = useDispatch();
```
and
```
onClick={() => dispatch(counterState.actions.increment())}
```