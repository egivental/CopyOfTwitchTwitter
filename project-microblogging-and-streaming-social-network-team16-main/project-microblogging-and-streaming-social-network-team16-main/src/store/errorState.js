import { createSlice } from "@reduxjs/toolkit";

const errorState = createSlice({
  name: "error",
  initialState: {
    errors: [],
  },
  reducers: {
    pushError: (state, action) => {
      state.errors.push(action.payload);
    },
    popError: (state) => {
      state.errors.shift();
    },
  },
});

export const { pushError, popError } = errorState.actions;

export default errorState;
