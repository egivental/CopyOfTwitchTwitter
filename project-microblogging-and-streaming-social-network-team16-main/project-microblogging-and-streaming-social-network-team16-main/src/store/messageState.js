import { createSlice } from "@reduxjs/toolkit";
import { apiBaseUrl } from "../constants/constants";
import { pushError } from "./errorState";

const messageState = createSlice({
  name: "message",
  initialState: {
    messages: null,
  },
  reducers: {
    updateMessages: (state, action) => {
      state.messages = action.payload;
    },
  },
});

export const { updateMessages } = messageState.actions;

export const getMessages = (username) => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/messages/${username}`, {
    method: "GET",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
  });
  const data = await response.json();
  if (response.status === 200) {
    await dispatch(updateMessages(data));
  } else {
    await dispatch(pushError(data.message));
  }
};

export const createMessage = (
  username,
  message,
  attachment,
  attachmentMimeType
) => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/messages/${username}`, {
    method: "POST",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      message,
      attachment,
      attachment_mime_type: attachmentMimeType,
    }),
  });
  if (response.status === 201) {
    await dispatch(getMessages(username));
  } else {
    const data = await response.json();
    await dispatch(pushError(data.message));
  }
};

export default messageState;
