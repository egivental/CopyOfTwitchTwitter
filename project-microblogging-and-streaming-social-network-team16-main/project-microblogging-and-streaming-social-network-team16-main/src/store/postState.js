import { createSlice } from "@reduxjs/toolkit";
import { apiBaseUrl } from "../constants/constants";
import { pushError } from "./errorState";

const postState = createSlice({
  name: "post",
  initialState: {
    posts: null,
  },
  reducers: {
    updatePosts: (state, action) => {
      state.posts = action.payload;
    },
  },
});

export const { updatePosts } = postState.actions;

export const getPosts = () => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/posts`, {
    method: "GET",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
  });
  const data = await response.json();
  if (response.status === 200) {
    await dispatch(updatePosts(data));
  } else {
    await dispatch(pushError(data.message));
  }
};

export const createPost = (message, attachment, attachmentMimeType) => async (
  dispatch
) => {
  const response = await fetch(`${apiBaseUrl}/post`, {
    method: "POST",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      message,
      attachment,
      attachment_mime_type: attachmentMimeType,
    }),
  });
  if (response.status === 201) {
    await dispatch(getPosts());
  } else {
    const data = await response.json();
    await dispatch(pushError(data.message));
  }
};

export const deletePost = (postId) => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/posts/${postId}`, {
    method: "DELETE",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
  });
  if (response.status === 200) {
    await dispatch(getPosts());
  } else {
    const data = await response.json();
    await dispatch(pushError(data.message));
  }
};

export const createComment = (postId, message) => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/posts/${postId}/comment`, {
    method: "POST",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      message,
    }),
  });
  if (response.status === 201) {
    await dispatch(getPosts());
  } else {
    const data = await response.json();
    await dispatch(pushError(data.message));
  }
};

export const updateComment = (postId, commentId, message) => async (
  dispatch
) => {
  const response = await fetch(
    `${apiBaseUrl}/posts/${postId}/comments/${commentId}`,
    {
      method: "PUT",
      credentials: "include",
      cache: "no-cache",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        message,
      }),
    }
  );
  if (response.status === 200) {
    await dispatch(getPosts());
  } else {
    const data = await response.json();
    await dispatch(pushError(data.message));
  }
};

export const deleteComment = (postId, commentId) => async (dispatch) => {
  const response = await fetch(
    `${apiBaseUrl}/posts/${postId}/comments/${commentId}`,
    {
      method: "DELETE",
      credentials: "include",
      cache: "no-cache",
      headers: {
        "Content-Type": "application/json",
      },
    }
  );
  if (response.status === 200) {
    await dispatch(getPosts());
  } else {
    const data = await response.json();
    await dispatch(pushError(data.message));
  }
};

export default postState;
