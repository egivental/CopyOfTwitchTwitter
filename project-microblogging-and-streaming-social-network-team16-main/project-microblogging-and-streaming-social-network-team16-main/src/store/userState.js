import { createSlice } from "@reduxjs/toolkit";
import { apiBaseUrl } from "../constants/constants";
import { pushError } from "./errorState";

const userState = createSlice({
  name: "user",
  initialState: {
    isUserLoggedIn: null,
    user: null,
    followers: [],
    followees: [],
    otherUsers: [],
  },
  reducers: {
    updateUser: (state, action) => {
      state.user = action.payload;
      if (state.user === null) {
        state.isUserLoggedIn = false;
        state.followers = [];
        state.followees = [];
        state.otherUsers = [];
      } else {
        state.isUserLoggedIn = true;
      }
    },
    updateFollowers: (state, action) => {
      state.followers = action.payload;
    },
    updateFollowees: (state, action) => {
      state.followees = action.payload;
    },
    updateOtherUsers: (state, action) => {
      state.otherUsers = action.payload;
    },
  },
});

export const {
  updateUser,
  updateFollowers,
  updateFollowees,
  updateOtherUsers,
} = userState.actions;

export const createUser = (username, password) => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/user`, {
    method: "POST",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      username,
      password,
      photo: null,
      photo_mime_type: null,
    }),
  });
  if (response.status === 201) {
    await dispatch(getSession());
  } else {
    const data = await response.json();
    await dispatch(pushError(data.message));
  }
};

export const getUsers = () => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/users`, {
    method: "GET",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
  });
  const data = await response.json();
  if (response.status === 200) {
    await dispatch(updateOtherUsers(data));
  } else {
    await dispatch(pushError(data.message));
  }
};

export const getUser = (username) => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/users/${username}`, {
    method: "GET",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
  });
  const data = await response.json();
  if (response.status === 200) {
    await dispatch(updateUser(data));
  } else {
    await dispatch(pushError(data.message));
    await dispatch(updateUser(null));
  }
};

export const updateUserProfile = (
  username,
  password,
  photo,
  photoMimeType
) => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/users/${username}`, {
    method: "PUT",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      password,
      photo,
      photo_mime_type: photoMimeType,
    }),
  });
  const data = await response.json();
  if (response.status === 200) {
    await dispatch(getUser(username));
    alert("User profile successfully updated!");
  } else {
    await dispatch(pushError(data.message));
  }
};

export const deleteUser = (username) => async (dispatch, getState) => {
  const state = getState();
  const loggedInUsername = state.user.user.username;
  const response = await fetch(`${apiBaseUrl}/users/${username}`, {
    method: "DELETE",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
  });
  const data = await response.json();
  if (response.status === 200) {
    if (username === loggedInUsername) {
      // Deactivated own account
      await dispatch(deleteSession());
    } else {
      // Blocked another user, refresh followers, followees, and other users
      await dispatch(getUsers());
      await dispatch(getFollowers(username));
      await dispatch(getFollowees(username));
    }
  } else {
    await dispatch(pushError(data.message));
  }
};

export const getFollowers = (username) => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/users/${username}/followers`, {
    method: "GET",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
  });
  const data = await response.json();
  if (response.status === 200) {
    await dispatch(updateFollowers(data));
  } else {
    await dispatch(pushError(data.message));
  }
};

export const getFollowees = (username) => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/users/${username}/followees`, {
    method: "GET",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
  });
  const data = await response.json();
  if (response.status === 200) {
    await dispatch(updateFollowees(data));
  } else {
    await dispatch(pushError(data.message));
  }
};

export const followUser = (username, followeeUsername) => async (dispatch) => {
  const response = await fetch(
    `${apiBaseUrl}/users/${username}/followees/${followeeUsername}`,
    {
      method: "PUT",
      credentials: "include",
      cache: "no-cache",
      headers: {
        "Content-Type": "application/json",
      },
    }
  );
  if (response.status === 200) {
    await dispatch(getFollowees(username));
    await dispatch(getUsers());
  } else {
    const data = await response.json();
    await dispatch(pushError(data.message));
  }
};

export const unfollowUser = (username, followeeUsername) => async (
  dispatch
) => {
  const response = await fetch(
    `${apiBaseUrl}/users/${username}/followees/${followeeUsername}`,
    {
      method: "DELETE",
      credentials: "include",
      cache: "no-cache",
      headers: {
        "Content-Type": "application/json",
      },
    }
  );
  if (response.status === 200) {
    await dispatch(getFollowees(username));
    await dispatch(getUsers());
  } else {
    const data = await response.json();
    await dispatch(pushError(data.message));
  }
};

export const getSession = () => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/session`, {
    method: "GET",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
  });
  const data = await response.json();
  if (response.status === 200) {
    await dispatch(getUser(data.username));
  } else {
    await dispatch(updateUser(null));
  }
};

export const createSession = (username, password) => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/session`, {
    method: "POST",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      username,
      password,
    }),
  });
  if (response.status === 201) {
    await dispatch(getUser(username));
  } else {
    const data = await response.json();
    await dispatch(pushError(data.message));
  }
};

export const deleteSession = () => async (dispatch) => {
  await fetch(`${apiBaseUrl}/session`, {
    method: "DELETE",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
  });
  await dispatch(getSession());
};

export const getMediaUrl = (id) => `${apiBaseUrl}/media/${id}`;

export default userState;
