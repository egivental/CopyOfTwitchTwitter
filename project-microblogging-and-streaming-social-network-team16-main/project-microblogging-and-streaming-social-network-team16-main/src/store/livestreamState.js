import { createSlice } from "@reduxjs/toolkit";
import { apiBaseUrl } from "../constants/constants";
import { pushError } from "./errorState";
import { getSession } from "./userState";

const livestreamState = createSlice({
  name: "livestream",
  initialState: {
    livestreams: null,
  },
  reducers: {
    updateLivestreams: (state, action) => {
      state.livestreams = action.payload;
    },
  },
});

export const { updateLivestreams } = livestreamState.actions;

export const createLivestream = () => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/livestream`, {
    method: "POST",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
  });
  if (response.status === 201) {
    await dispatch(getLivestreams());
  } else {
    const data = await response.json();
    await dispatch(pushError(data.message));
  }
};

export const getLivestreams = () => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/livestreams`, {
    method: "GET",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
  });
  const data = await response.json();
  if (response.status === 200) {
    await dispatch(updateLivestreams(data));
    await dispatch(getSession());
  } else {
    await dispatch(pushError(data.message));
  }
};

export const deleteLivestream = (room) => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/livestreams/${room}`, {
    method: "DELETE",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
  });
  if (response.status === 200) {
    await dispatch(getLivestreams());
  } else {
    const data = await response.json();
    await dispatch(pushError(data.message));
  }
};

export const createLivestreamComment = (room, message) => async (dispatch) => {
  const response = await fetch(`${apiBaseUrl}/livestreams/${room}/comment`, {
    method: "POST",
    credentials: "include",
    cache: "no-cache",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      message,
    }),
  });
  if (response.status === 201) {
    await dispatch(getLivestreams());
  } else {
    const data = await response.json();
    await dispatch(pushError(data.message));
  }
};

export const updateLivestreamComment = (room, commentId, message) => async (
  dispatch
) => {
  const response = await fetch(
    `${apiBaseUrl}/livestreams/${room}/comments/${commentId}`,
    {
      method: "PUT",
      credentials: "include",
      cache: "no-cache",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        message,
      }),
    }
  );
  if (response.status === 200) {
    await dispatch(getLivestreams());
  } else {
    const data = await response.json();
    await dispatch(pushError(data.message));
  }
};

export const deleteLivestreamComment = (room, commentId) => async (
  dispatch
) => {
  const response = await fetch(
    `${apiBaseUrl}/livestreams/${room}/comments/${commentId}`,
    {
      method: "DELETE",
      credentials: "include",
      cache: "no-cache",
      headers: {
        "Content-Type": "application/json",
      },
    }
  );
  if (response.status === 200) {
    await dispatch(getLivestreams());
  } else {
    const data = await response.json();
    await dispatch(pushError(data.message));
  }
};

export default livestreamState;
