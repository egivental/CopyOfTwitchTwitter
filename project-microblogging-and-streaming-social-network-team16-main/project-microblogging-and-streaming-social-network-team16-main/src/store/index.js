import { configureStore } from "@reduxjs/toolkit";
import errorState from "./errorState";
import messageState from "./messageState";
import postState from "./postState";
import userState from "./userState";
import livestreamState from "./livestreamState";

export default configureStore({
  reducer: {
    error: errorState.reducer,
    user: userState.reducer,
    post: postState.reducer,
    message: messageState.reducer,
    livestream: livestreamState.reducer,
  },
});
