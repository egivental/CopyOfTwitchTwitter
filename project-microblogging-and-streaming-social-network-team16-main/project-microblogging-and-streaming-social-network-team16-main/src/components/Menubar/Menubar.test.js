import React from "react";
import { render } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import store from "../../store/index";
import Menubar from "./Menubar";

test("renders page", () => {
  const { getByText } = render(
    <Provider store={store}>
      <Menubar />
    </Provider>
  );

  expect(getByText(/ABE Twitter/i)).toBeInTheDocument();
  expect(getByText(/Sign In/i)).toBeInTheDocument();
  expect(getByText(/Sign Up/i)).toBeInTheDocument();
});

// snapshot testing
test("Menubar matches snapshot", () => {
  const component = renderer.create(
    <Provider store={store}>
      <Menubar />
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
