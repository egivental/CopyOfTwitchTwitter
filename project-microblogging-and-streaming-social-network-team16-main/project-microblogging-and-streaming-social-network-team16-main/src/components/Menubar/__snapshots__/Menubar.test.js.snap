// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menubar matches snapshot 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-light bg-light"
>
  <a
    className="navbar-brand"
    href="/home"
  >
    <strong>
      ABE Twitter
    </strong>
  </a>
  <button
    aria-label="Toggle navigation"
    className="navbar-toggler collapsed"
    onClick={[Function]}
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <div
    aria-expanded={null}
    className="navbar-collapse collapse"
    id="basic-navbar-nav"
  >
    <div
      className="ml-auto navbar-nav"
      onKeyDown={[Function]}
    >
      <a
        className="nav-link"
        data-rb-event-key="/signup"
        href="/signup"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        Sign Up
      </a>
      <a
        className="nav-link"
        data-rb-event-key="/signin"
        href="/signin"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        Sign In
      </a>
    </div>
  </div>
</nav>
`;
